src/libos/crates/host-socket/src/datagram/netlink.rs

```diff
@@ -0,0 +1,256 @@
+use super::*;
+use async_io::socket::{NetlinkFamily, Type};
+use byteorder::{ByteOrder, NativeEndian};
+use core::ops::{Range, RangeFrom};
+
+pub struct NetlinkSocket<A: Addr + 'static, R: Runtime> {
+    common: Arc<Common<A, R>>,
+    sender: Sender<A, R>,
+    receiver: Arc<Receiver<A, R>>,
+}
+
+impl<A: Addr, R: Runtime> NetlinkSocket<A, R> {
+    pub fn new(
+        socket_type: Type,
+        netlink_family: NetlinkFamily,
+        nonblocking: bool,
+    ) -> Result<Self> {
+        let common = Arc::new(Common::new(
+            socket_type,
+            nonblocking,
+            Some(netlink_family as i32),
+        )?);
+        let sender = Sender::new(common.clone());
+        let receiver = Receiver::new(common.clone());
+
+        // Start async recv
+        receiver.initiate_async_recv();
+
+        Ok(Self {
+            common,
+            sender,
+            receiver,
+        })
+    }
+
+    pub fn domain(&self) -> Domain {
+        A::domain()
+    }
+
+    pub fn host_fd(&self) -> HostFd {
+        self.common.host_fd()
+    }
+
+    pub fn status_flags(&self) -> StatusFlags {
+        // Only support O_NONBLOCK
+        if self.common.nonblocking() {
+            StatusFlags::O_NONBLOCK
+        } else {
+            StatusFlags::empty()
+        }
+    }
+
+    pub fn set_status_flags(&self, new_flags: StatusFlags) -> Result<()> {
+        // Only support O_NONBLOCK
+        let nonblocking = new_flags.is_nonblocking();
+        self.common.set_nonblocking(nonblocking);
+        Ok(())
+    }
+
+    pub fn addr(&self) -> Result<A> {
+        let common = &self.common;
+
+        // Always get addr from host.
+        // Because for IP socket, users can specify "0" as port and the kernel should select a usable port for him.
+        // Thus, when calling getsockname, this should be updated.
+        let addr = common.get_addr_from_host()?;
+        common.set_addr(&addr);
+        Ok(addr)
+    }
+
+    pub fn peer_addr(&self) -> Result<A> {
+        Ok(self.common.peer_addr().unwrap_or_default())
+    }
+
+    pub async fn connect(&self, peer_addr: Option<&A>) -> Result<()> {
+        if let Some(peer) = peer_addr {
+            self.common.set_peer_addr(peer);
+            // Start async recv after explicit binding or implicit binding
+            self.receiver.initiate_async_recv();
```

Comment:
In the `NetlinkSocket::new()`, `init_async_recv` has been called, is it necessary here?